internal/gadget-runner/gadgetrunner.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2022-2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+Package gadgetrunner handles initializing gadgets and installed operators before
+handing them over to a specified runtime.
+*/
+package gadgetrunner
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/inspektor-gadget/inspektor-gadget/internal/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/parser"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/runtime"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+// GadgetRunner handles running gadgets by the gadget interface; it orchestrates the whole lifecycle of the gadget
+// instance and communicates with gadget and runtime.
+type GadgetRunner struct {
+	ctx          context.Context
+	id           string
+	gadget       gadgets.Gadget
+	gadgetParams *params.Params
+	runtime      runtime.Runtime
+	parser       parser.Parser
+	operators    operators.Operators
+	logger       logger.Logger
+	result       []byte
+	resultError  error
+}
+
+func NewGadgetRunner(
+	ctx context.Context,
+	id string,
+	runtime runtime.Runtime,
+	gadget gadgets.Gadget,
+	parser parser.Parser,
+	logger logger.Logger,
+) *GadgetRunner {
+	return &GadgetRunner{
+		ctx:       ctx,
+		id:        id,
+		gadget:    gadget,
+		runtime:   runtime,
+		parser:    parser,
+		logger:    logger,
+		operators: operators.GetOperatorsForGadget(gadget),
```

Comment:
It seems the caller of `NewGadgetRunner()` cannot choose the list of operators: it is based on what operators are compatible and are compiled in and call `operators.Register()` in an `init()` function. It seems wrong to me: what if the localmanager is compiled it but we want to use something else? For example, we might have localmanager to filter by containers and systemdmanager to filter by systemd units. The user would select the filter on either systemd units or container names, but both operators should probably not be used, shouldn't they?

Reply:
If the list of operators was more explicit, we could get a reference to them and it might be easier to set the parameters in a structured way without using strings (@mauriciovasquezbernal 's concern at https://github.com/inspektor-gadget/inspektor-gadget/pull/1281#issuecomment-1409163064)

Reply:
Yeah, it makes sense to be able to choose from - maybe we could add something like --disable-operators=LocalManager for that. Operators also can already have params to disable them by themselves, if needed. For example a param '--no-local-manager' could be added to LocalManager and evaluated before actually running / doing its work. If you want to include mutually exclusive options, we could implement that by for example using tags - so that for each tag only one operator can work at a time. But I think we can add those things later on, right?

These kind of exceptions have to be extended anyways, because the checks for which operators can be used for what gadget should run on the server (or be baked in information, not totally sure which way to go) when using kubectl-gadget (as there gadgets cannot be instantiated and 'CanOperateOn' checks could fail for operators because of that). kubectl-gadget still would have to know about that to ask for the params.